### PR TITLE
doc: add instructions to remove a package

### PR DESCRIPTION
**Pull request by davidak**
###### Motivation for this change

It was not documented and i had to reverse engineer it.

It don't really fit in that place and the style is different than the rest. I wish every task would be explained that detailed, so everyone can do it.

## Preview

https://github.com/NixOS/nixpkgs/pull/116475/files?short_path=ba88122#diff-ba881223dddc816723e79c2ee074df7b18a08b6520273c22665143b700f3f52d

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
FYI I can just press that button for a preview.

image

File: doc/contributing/submitting-changes.chapter.md [hunk 1]
```diff
@@ -82,6 +82,48 @@ If a security fix applies to both master and a stable release then, similar to r
 
 Critical security fixes may by-pass the staging branches and be delivered directly to release branches such as `master` and `release-*`.
 
+## Deprecating/removing packages {#submitting-changes-deprecating-packages}
+
+When software projects get discontinued or are unmaintained and break, we have to remove them. Another reason would be that there is no maintainer for the package in nixpkgs. In that case, try to find a new maintainer first.
```
Comment (SuperSandro2000) [inline doc/contributing/submitting-changes.chapter.md:87]:
```suggestion
When software projects got discontinued, are unmaintained, broken for a long time or have severe security vulnerabilities, we have the option to remove them. Another reason would be that for a longer time there is no maintainer for the package in nixpkgs. In any case, it should always be attempted to find a new maintainer or patch smaller issues first.
```

Maybe we could add examples when a package can definitely be removed or not.

Can be safely removed:
- Package is broken since one or several years and did not receive meaningful commits (commits that are not treewide changes.).
- ...

Can not be removed:
- broken since last week due to an unrelated change.
- ...

File: doc/contributing/submitting-changes.chapter.md [hunk 2]
```diff
@@ -82,6 +82,48 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+### Steps to remove a package from nixpkgs
+
+We use jbidwatcher as an example for a discontinued project here.
+
+1. have nixpkgs checked out locally and up to date
+1. create a new branch for your change, e.g. `git checkout -b jbidwatcher`
+1. remove the actual package including it's directory, e.g. `rm -rf pkgs/applications/misc/jbidwatcher`
+1. remove the package from list of all packages (`pkgs/top-level/all-packages.nix`)
+1. add error message for package name in `pkgs/top-level/aliases.nix`
```
Comment (SuperSandro2000) [inline doc/contributing/submitting-changes.chapter.md:97]:
```suggestion
1. add an alias for the package name in `pkgs/top-level/aliases.nix`
```

For python this will be most likely python-aliases and vim has also its own aliases file.

File: doc/contributing/submitting-changes.chapter.md [hunk 3]
```diff
@@ -82,6 +82,48 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+    For example in this case:
+    ```
+    jbidwatcher = throw "jbidwatcher was discontinued in march 2021"; # added 2021-03-15
```
Comment (SuperSandro2000) [inline doc/contributing/submitting-changes.chapter.md:101]:
```suggestion
    jbidwatcher = throw "jbidwatcher was discontinued in march 2020"; # added 2021-03-15
```

I don't know if this is a real world example but using a earlier date would emphasize that the package is old and stale.

File: doc/contributing/submitting-changes.chapter.md [hunk 4]
```diff
@@ -82,6 +82,48 @@ If a security fix applies to both master and a stable release then, similar to r
...
+    ```
+
+    Explain shortly why the package was removed for users that still have it installed.
```
Comment (SuperSandro2000) [inline doc/contributing/submitting-changes.chapter.md:104]:
```suggestion
    The throw message should explain in short why the package was removed for users that still have it installed.
```

File: doc/contributing/submitting-changes.chapter.md [hunk 5]
```diff
@@ -82,6 +82,48 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+1. test if the change introduced any issues with `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
```
Comment (SuperSandro2000) [inline doc/contributing/submitting-changes.chapter.md:106]: ``nix-env -qaP -f . --show-trace`` with aliases off is enough. Not sure how to pass ``config.allowAliases = false;`` to it.

**Review by SuperSandro2000 — COMMENTED** [review]
Great idea!

I think grammatical correct sentences need to start with a capital letter.

File: doc/contributing/submitting-changes.chapter.md (hunk 3 above)
Comment (davidak) [reply doc/contributing/submitting-changes.chapter.md:101]: it is an actual example which you just merged :smile:

Comment (SuperSandro2000) [reply doc/contributing/submitting-changes.chapter.md:101]: That was after I wrote this.

File: doc/contributing/submitting-changes.chapter.md (hunk 5 above)
Comment (davidak) [reply doc/contributing/submitting-changes.chapter.md:106]: Yeah, that's better. nixpkgs-review does not even work in this case :grimacing: 

https://github.com/Mic92/nixpkgs-review/issues/189

Comment (SuperSandro2000) [reply doc/contributing/submitting-changes.chapter.md:106]: I run it locally with that option since a bit but taking a look at the issue.

**Comment by davidak** [comment]
I finally found the time to update it. Feel free to give feedback!

**Comment by davidak** [comment]
@FRidh where is the right place to add python aliases?

**Comment by FRidh** [comment]
Package sets typically do not have aliases.

File: doc/contributing/submitting-changes.chapter.md [hunk 6]
```diff
@@ -82,6 +82,50 @@ If a security fix applies to both master and a stable release then, similar to r
 
 Critical security fixes may by-pass the staging branches and be delivered directly to release branches such as `master` and `release-*`.
 
+## Deprecating/removing packages {#submitting-changes-deprecating-packages}
+
+When software projects got discontinued, are unmaintained for a year, broken for a year or have severe security vulnerabilities that don't get fixed, we should remove them. Another reason would be that there is no maintainer for the package in nixpkgs for more than 6 months. In any case, it should always be attempted to find a new maintainer or fix smaller issues first.
```
Comment (FRidh) [inline doc/contributing/submitting-changes.chapter.md:87]: > When software projects got discontinued, are unmaintained for a year, broken for a year or have severe security vulnerabilities that don't get fixed, we should remove them. Another reason would be that there is no maintainer for the package in nixpkgs for more than 6 months.

This is quite a step and controversial. Probably better to leave the "when to remove" and stick with "how to remove" in this PR.

File: doc/contributing/submitting-changes.chapter.md [hunk 7]
```diff
@@ -82,6 +82,50 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+### Steps to remove a package from nixpkgs
+
+We use jbidwatcher as an example for a discontinued project here.
+
+1. have nixpkgs checked out locally and up to date
+1. create a new branch for your change, e.g. `git checkout -b jbidwatcher`
+1. remove the actual package including it's directory, e.g. `rm -rf pkgs/applications/misc/jbidwatcher`
+1. remove the package from list of all packages (`pkgs/top-level/all-packages.nix`)
+1. add an alias for the package name in `pkgs/top-level/aliases.nix` (there is also `pkgs/misc/vim-plugins/aliases.nix` and `pkgs/top-level/python-packages.nix` for python packages)
```
Comment (FRidh) [inline doc/contributing/submitting-changes.chapter.md:97]: `python-packages.nix` should not keep aliases. There are some, but those should be cleaned up. A separate aliases file may be created for this functionality but does not exist yet.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: doc/contributing/submitting-changes.chapter.md (hunk 6 above)
Comment (davidak) [reply doc/contributing/submitting-changes.chapter.md:87]: How can we come to a policy here? Discuss in Forum or does it need an RFC?

File: doc/contributing/submitting-changes.chapter.md (hunk 7 above)
Comment (davidak) [reply doc/contributing/submitting-changes.chapter.md:97]: OK

**Comment by SuperSandro2000** [comment]
> @FRidh where is the right place to add python aliases?

Not merged yet.

**Comment by davidak** [comment]
Can we get this merged?

@SuperSandro2000 @FRidh

**Comment by davidak** [comment]
Can you add your review again?

File: doc/contributing/submitting-changes.chapter.md [hunk 8]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
 
 Critical security fixes may by-pass the staging branches and be delivered directly to release branches such as `master` and `release-*`.
 
+## Deprecating/removing packages {#submitting-changes-deprecating-packages}
+
+There is currently no policy when to remove a package.
+
+It should always be attempted to find a new maintainer or fix smaller issues first.
+
+### Steps to remove a package from nixpkgs
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:91]:
```suggestion
### Steps to remove a package from Nixpkgs
```

File: doc/contributing/submitting-changes.chapter.md [hunk 9]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+We use jbidwatcher as an example for a discontinued project here.
+
+1. Have nixpkgs checked out locally and up to date.
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:95]:
```suggestion
1. Have Nixpkgs checked out locally and up to date.
```

File: doc/contributing/submitting-changes.chapter.md [hunk 10]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Create a new branch for your change, e.g. `git checkout -b jbidwatcher`
+1. Remove the actual package including it's directory, e.g. `rm -rf pkgs/applications/misc/jbidwatcher`
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:97]:
```suggestion
1. Remove the actual package including its directory, e.g. `rm -rf pkgs/applications/misc/jbidwatcher`
```

File: doc/contributing/submitting-changes.chapter.md [hunk 11]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Remove the package from list of all packages (`pkgs/top-level/all-packages.nix`).
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:98]:
```suggestion
1. Remove the package from the list of all packages (`pkgs/top-level/all-packages.nix`).
```

File: doc/contributing/submitting-changes.chapter.md [hunk 12]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+    For example in this case:
+    ```
+    jbidwatcher = throw "jbidwatcher was discontinued in march 2021"; # added 2021-03-15
+    ```
+
+    The throw message should explain in short why the package was removed for users that still have it installed.
+
+1. Test if the changes introduced any issues with `nix-env -qaP -f . --show-trace`. It should show the list of packages and not an error.
+1. Commit the changes. Explain again why the package was removed. When it was declared discontinued upstream, add a link to the source.
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:109]:
```suggestion
1. Commit the changes. Explain again why the package was removed. If it was declared discontinued upstream, add a link to the source.
```

File: doc/contributing/submitting-changes.chapter.md [hunk 13]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Add an alias for the package name in `pkgs/top-level/aliases.nix` (There is also `pkgs/misc/vim-plugins/aliases.nix`. Package sets typically do not have aliases, so we can't add them there.)
+
+    For example in this case:
+    ```
+    jbidwatcher = throw "jbidwatcher was discontinued in march 2021"; # added 2021-03-15
+    ```
+
+    The throw message should explain in short why the package was removed for users that still have it installed.
+
+1. Test if the changes introduced any issues with `nix-env -qaP -f . --show-trace`. It should show the list of packages and not an error.
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:108]:
```suggestion
1. Test if the changes introduced any issues by running `nix-env -qaP -f . --show-trace`. It should show the list of packages without errors.
```

File: doc/contributing/submitting-changes.chapter.md [hunk 14]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Commit the changes. Explain again why the package was removed. When it was declared discontinued upstream, add a link to the source.
+
+    ```
+    git add pkgs/applications/misc/jbidwatcher/default.nix pkgs/top-level/all-packages.nix pkgs/top-level/aliases.nix
+    git commit
+    ```
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:114]:
```suggestion
    ```ShellSession
    $ git add pkgs/applications/misc/jbidwatcher/default.nix pkgs/top-level/all-packages.nix pkgs/top-level/aliases.nix
    $ git commit
    ```
```

File: doc/contributing/submitting-changes.chapter.md [hunk 15]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+There is currently no policy when to remove a package.
+
+It should always be attempted to find a new maintainer or fix smaller issues first.
+
+### Steps to remove a package from nixpkgs
+
+We use jbidwatcher as an example for a discontinued project here.
+
+1. Have nixpkgs checked out locally and up to date.
+1. Create a new branch for your change, e.g. `git checkout -b jbidwatcher`
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:96]:
```suggestion
1. Create a new branch for your change, e.g. `git checkout -b jbidwatcher`
```

File: doc/contributing/submitting-changes.chapter.md [hunk 16]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Remove the actual package including it's directory, e.g. `rm -rf pkgs/applications/misc/jbidwatcher`
+1. Remove the package from list of all packages (`pkgs/top-level/all-packages.nix`).
+1. Add an alias for the package name in `pkgs/top-level/aliases.nix` (There is also `pkgs/misc/vim-plugins/aliases.nix`. Package sets typically do not have aliases, so we can't add them there.)
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:99]:
```suggestion
1. Add an alias for the package name in `pkgs/top-level/aliases.nix` (There is also `pkgs/misc/vim-plugins/aliases.nix`. Package sets typically do not have aliases, so we can't add them there.)
```

File: doc/contributing/submitting-changes.chapter.md [hunk 17]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+1. Test if the changes introduced any issues with `nix-env -qaP -f . --show-trace`. It should show the list of packages and not an error.
+1. Commit the changes. Explain again why the package was removed. When it was declared discontinued upstream, add a link to the source.
+
+    ```
+    git add pkgs/applications/misc/jbidwatcher/default.nix pkgs/top-level/all-packages.nix pkgs/top-level/aliases.nix
+    git commit
+    ```
+
+    commit message:
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:116]:
```suggestion
    Example commit message:
```

File: doc/contributing/submitting-changes.chapter.md [hunk 18]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+
+    ```
+    jbidwatcher: remove
+
+    project was discontinued in march 2021. the program does not work anymore because ebay changed the login.
+
+    https://web.archive.org/web/20210315205723/http://www.jbidwatcher.com/
+    ```
+
+1. Push changes to your GitHub fork with `git push`
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:126]:
```suggestion
1. Push changes to your GitHub fork with `git push`.
```

File: doc/contributing/submitting-changes.chapter.md [hunk 19]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
...
+1. Create a pull request against nixpkgs. Mention the package maintainer.
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:127]:
```suggestion
1. Create a pull request against Nixpkgs. Mention the package maintainer.
```

**Review by siraben — CHANGES_REQUESTED** [review]
Overall structure looks OK. Suggested grammatical/phrasing fixes. Not sure if the list numbering staying at `1.` was intentional or not.

EDIT: list numbering is fine on export

File: doc/contributing/submitting-changes.chapter.md [hunk 20]
```diff
@@ -82,6 +82,52 @@ If a security fix applies to both master and a stable release then, similar to r
 
 Critical security fixes may by-pass the staging branches and be delivered directly to release branches such as `master` and `release-*`.
 
+## Deprecating/removing packages {#submitting-changes-deprecating-packages}
+
+There is currently no policy when to remove a package.
+
+It should always be attempted to find a new maintainer or fix smaller issues first.
```
Comment (siraben) [inline doc/contributing/submitting-changes.chapter.md:89]: Before removing a package, one should try to find a new maintainer or fix smaller issues first.

File: doc/contributing/submitting-changes.chapter.md (hunk 18 above)
Comment (davidak) [reply doc/contributing/submitting-changes.chapter.md:126]: this would make it harder to select and copy
with the command on the right, you can just select from right

File: doc/contributing/submitting-changes.chapter.md (hunk 20 above)
Comment (davidak) [inline doc/contributing/submitting-changes.chapter.md:89]:
```suggestion
Before removing a package, one should try to find a new maintainer or fix smaller issues first.
```

**Comment by davidak** [comment]
@siraben thanks. i accepted the suggestions except one. take a look if you agree that this makes sense

i also squashed, so it could get merged now

File: doc/contributing/submitting-changes.chapter.md (hunk 18 above)
Comment (siraben) [reply doc/contributing/submitting-changes.chapter.md:126]: Good point

**Review by samueldr — APPROVED** [review]

**Review by siraben — APPROVED** [review]
